### PR TITLE
Enhanced preview and splitting

### PR DESCRIPTION
**Pull request by Recidiviste**
Hello there !

Add more options to preview, like syntax highlighting, archives content preview, pdf preview....
All of that only if the packages necessary are present (so same dependencies).
Plus, printing more informations using `stat` about the file if the type is unknown.
Finally, ask for the type of splitting when in Tmux

PS : I'm not a big shell writer, so if it's better to have case instead of if-elif-then block, I can change that. You have the editings right on this after all, so change it the way you want. I love your project.

### CODE REVIEW

**Comment by jarun** [comment]
Hi, thanks for the PR.

@leovilok can you please take a look? I have a small concern about extracting the archives and  listing them. Do you think it will slow the previewer down?

**Comment by jarun** [comment]
Also, @Recidiviste CircleCI fails... probably due to `shellcheck`. Please take a look.

**Comment by Recidiviste** [comment]
Done, I forgetted the `"` around some tests and also that `$lines` did disappeared between the version I was working on and the last version (sorry for being an amateur)

File: plugins/preview-tui [hunk 1]
```diff
@@ -74,7 +94,16 @@ if [ "$PREVIEW_MODE" ] ; then
 fi
 
 if [ -e "${TMUX%%,*}" ] && [ "$(tmux -V | cut -c6)" -eq 3 ] ; then
-    tmux split-window -e "NNN_FIFO=$NNN_FIFO" -e "PREVIEW_MODE=1" -dh "$0" "$1"
+   printf "split 'h'orizontal/'v'ertical :"
```
Comment (jarun) [inline plugins/preview-tui:97]: @leovilok I think we discussed this at some point. We either have a variable or do this by determine the rows*cols. Prompting everything is not very user-friendly.

Comment (Recidiviste) [reply plugins/preview-tui:97]: Maybe if there's more cols than (lines * 2), split vertically (n³ | preview), else, split horizontally (n³ above ― preview below). The only problem is if cols and (lines * 2) are really close (for example 116 cols and 59 lines => easier with horizontal split, but vertical split using our condition).
Maybe add a variable ASK at the start, by default on false. If false, split using condition (or maybe set a default vertical split). If true, ask the type of split.
Writing that, I think it's getting more and more complex for a simple plugin. I suggest doing a first merge request for the different previews and a second for the type of split. Your choice.

Comment (jarun) [reply plugins/preview-tui:97]: I think we should do the following (tmux-only):

have a variable: SPLIT=h
this overrides all. if user sets to v than do vertical split

now, if the above is not set:
if lines > cols: do horizontal split
if cols > lines: do vertical split
if cols=lines: do vertical split

Comment (Recidiviste) [reply plugins/preview-tui:97]: Ok with that at the exception of the conditions lines > cols/lines < cols
On a terminal, there's way more cols than lines usually. Maybe add a multiplier (2 for example, even if it's not quite right) for the number of lines. Than we have

if (lines*2) > cols : do horizontal split
if (lines*2) < cols : do vertical split
if (lines*2) = cols : do vertical split

As said, it's your choice, it's not a big change in the code

Comment (jarun) [reply plugins/preview-tui:97]: OK. Let's try this.

**Comment by Recidiviste** [comment]
Now it works. Don't look at the feed up there, I didn't struggled at all doing this simple thing...

**Comment by leovilok** [comment]
Hi @Recidiviste and thanks for the PR, it's good to know that you could hack on `preview-tui`, plugins are made to be customisable after all.

I do have a few issues/comments on this PR:

1. You actually don't need the highlight call to have syntax highlighted text in less: you can configure less with the LESSOPEN env var to set an input filter (using the `highlight` command as LESSOPEN is actually a popular option)
2. (maybe the biggest issue) You cant really use lines like
    ~~~sh
    cmd | $PAGER &
    ~~~
   because the whole pipeline is one unique job in the shell, with `cmd`'s PID as the job's PID, so when you kill the job (because there is a new input) it actually only kills `cmd` and not `less`. With your version of the plugin, try highlighting several files with various filetypes, than on one file try scrolling in less: other less instance are still running so it will scroll a random less instance, showing any of the files you were previewing before. That's why I'm using a FIFO for `tree` view, we could work on a way to make it generic for any command if you want. (I hope that this explanation makes sense, j'aurais peut-être du te la faire en français :smile:)
3. I think that we should frefer `file -b --mime-type` over just `file -b`: the output should be more consistent while retaining the same info.
4. For archive preview, maybe `atool`'s `als` is a more universal tool? that's jut a matter of taste. @jarun with the async architecture it can't slow down anything, if it's not fast enough to list archive content before another file is highlighted... it just gets killed.
5. About viu/pdftotext: it's good to be able to have these options, but I'm feeling like we're reimplementing `nuke` plugin again... which is why I'm calling `nuke` directly in `preview-tabbed`. Maybe we could come up with a way to call `nuke` and still reliably kill what it spawns... there might be something to try there.
6. On tmux vertical/horizontal split: this code could use a bit of factorization, but I see that you autodetect width/height ratio which was a request from @jarun that I totally forgot about :smile:. I'm not sure about the SPLIT env var though, isn't there a Tmux configuration controlling the default split orientation?

I'm not sure that we'll get everything you added in `preview-tui` (it's supposed to be a barebone preview plugin after all), but there is definitively some good stuff there. And I hope that you had fun hacking it!

File: plugins/preview-tui [hunk 2]
```diff
@@ -23,14 +23,17 @@
 
 TERMINAL="${TERMINAL:-xterm}"
 PAGER="${PAGER:-less -r}"
+SPLIT="h"
```
Comment (jarun) [inline plugins/preview-tui:26]: Please keep it empty in the beginning with comment.

Also, please document this behaviour in the Usage section.

File: plugins/preview-tui [hunk 3]
```diff
@@ -47,13 +52,31 @@ preview_file () {
...
+     	pdftotext "$1" - | $PAGER &
+     elif ( [ "$typ" = "PNG" ] || [ "$typ" = "JPEG" ] ) && [ "$(command -v viu)" ]; then
+     	# Display image in terminal (more or less lowered resolution)
+     	viu "$1" | head -n "$lines"
     elif [ "$encoding" = "binary" ] ; then
         # Binary file: just print filetype info
-        echo "-------- Binary file --------"
+        echo '-------- Binary file --------'
         file -b "$1"
+	echo '------- Informations -------'
```
Comment (jarun) [inline plugins/preview-tui:71]: Informations -> Details

**Comment by jarun** [comment]
> I'm not sure about the SPLIT env var though, isn't there a Tmux configuration controlling the default split orientation?

@Recidiviste in that case just drop split.

File: plugins/preview-tui (hunk 2 above)
Comment (jarun) [reply plugins/preview-tui:26]: Please drop split. Probably add a note on the tmux variable.

Comment (leovilok) [reply plugins/preview-tui:26]: I was wrong, default layout can't be easily configured in tmux...

Comment (jarun) [reply plugins/preview-tui:26]: OK, then:



Please keep it empty in the beginning with comment.

Also, please document this behaviour in the Usage section.

File: plugins/preview-tui (hunk 3 above)
Comment (Recidiviste) [reply plugins/preview-tui:71]: Done

Comment (jarun) [reply plugins/preview-tui:71]: I meant:

    	echo '------- Details -------'

File: plugins/preview-tui [hunk 4]
```diff
@@ -47,13 +52,31 @@ preview_file () {
         )
 
         rm "$tmpfifopath"
+     elif [ "$typ" = "Zip" ] && [ "$(command -v zip)" ]; then
```
Comment (jarun) [inline plugins/preview-tui:55]: The alignment looks broken to me. Can you please check?

Comment (Recidiviste) [reply plugins/preview-tui:55]: Don't seems to me, 3 spaces like the rest of the file.

File: plugins/preview-tui [hunk 5]
```diff
@@ -47,13 +52,31 @@ preview_file () {
...
+     	# Display content of Zip
```
Comment (jarun) [inline plugins/preview-tui:56]: This line is not aligned correctly. Basically all statements after each condition check.

Comment (jarun) [reply plugins/preview-tui:56]: There's a TAB.

File: plugins/preview-tui (hunk 4 above)
Comment (jarun) [reply plugins/preview-tui:55]: It has 1 space extra wrt. line 38. Fix it please.

File: plugins/preview-tui (hunk 5 above)
Comment (jarun) [reply plugins/preview-tui:56]: We do not use TABs in the scripts.

File: plugins/preview-tui [hunk 6]
```diff
@@ -47,13 +56,31 @@ preview_file () {
...
-        echo "-------- Binary file --------"
+        echo '-------- Binary file --------'
         file -b "$1"
+	echo '------- Details -------'
+	stat "$1"
     else
         # Text file:
-        $PAGER "$1" &
+	if [ "$(command -v highlight)" ]; then
+		highlight --out-format=ansi --force "$1" | $PAGER &
```
Comment (jarun) [inline plugins/preview-tui:80]: And what's this alignment? I can see the differences in the GitHub diff. Please take a look yourself.

**Comment by jarun** [comment]
Also, CI fails.

@KlzXS and @leovilok is there a way to enforce 4 space alignment using CI? I cannot spend time on this trivia.

File: plugins/preview-tui [hunk 7]
```diff
@@ -47,13 +56,31 @@ preview_file () {
...
+    	pdftotext "$1" - | $PAGER &
+    elif ( [ "$typ" = "PNG" ] || [ "$typ" = "JPEG" ] ) && [ "$(command -v viu)" ]; then
+     	# Display image in terminal (more or less lowered resolution)
+     	viu "$1" | head -n "$lines"
     elif [ "$encoding" = "binary" ] ; then
         # Binary file: just print filetype info
-        echo "-------- Binary file --------"
+        echo '-------- Binary file --------'
         file -b "$1"
+	echo '------- Details -------'
```
Comment (jarun) [inline plugins/preview-tui:75]: You can see the difference between line 74 and 75.

File: plugins/preview-tui [hunk 8]
```diff
@@ -47,13 +56,31 @@ preview_file () {
         )
 
         rm "$tmpfifopath"
+    elif [ "$typ" = "Zip" ] && [ "$(command -v zip)" ]; then
+    	# Display content of Zip
```
Comment (jarun) [inline plugins/preview-tui:60]: This line should be aligned to 4 spaces.

Comment (jarun) [reply plugins/preview-tui:60]: So 4 + 4 = 8 spaces from the left.

File: plugins/preview-tui [hunk 9]
```diff
@@ -47,13 +56,31 @@ preview_file () {
...
         # Binary file: just print filetype info
-        echo "-------- Binary file --------"
+        echo '-------- Binary file --------'
         file -b "$1"
+	echo '------- Details -------'
+	stat "$1"
     else
         # Text file:
-        $PAGER "$1" &
+	if [ "$(command -v highlight)" ]; then
```
Comment (jarun) [inline plugins/preview-tui:79]: See line 78 and 79. I think your editor settings is the problem.

File: plugins/preview-tui [hunk 10]
```diff
@@ -18,17 +18,26 @@
...
+
+lines=$(($(tput lines)-1))
+cols=$(tput cols)
 
 preview_file () {
     kill "$(jobs -p)" 2>/dev/null
     clear
 
     encoding="$(file -b --mime-encoding "$1")"
+    typ="$(file -b --mime-type)"
```
Comment (jarun) [inline plugins/preview-tui:39]: For me:

file -b --mime-type abc.pdf 
application/pdf

Will it work? Also, where are you passing the file name?

Comment (leovilok) [reply plugins/preview-tui:39]: The output from `file -b --mime-type` is not the same as `file -b`, so further check need to be adapted. But as MIME types are standards, they should be the same across different systems with different implementations of `file`.

**Comment by jarun** [comment]
I just tested it in the current status and I noticed that the newly created pane just flickers and exits. No preview is seen. However, I can see that the splitting logic is in play (e.g. in my case with tmux it now tries to split horizontally).

@leovilok can you give it a try and see what all needs to be fixed? Probably we can pull it in and you can sanitize it in another iteration.

**Comment by jarun** [comment]
@Recidiviste I have pushed partial changes form your PR at commit 5df231947392475ced257a6ba5e4cec3c3fbb5a9. This covers the terminal splitting logic.

I think you can re-raise the PR with only the mime related changes and the alignment issues fixed.

**Comment by jarun** [comment]
@Recidiviste In subsequent commits ca8bc8bf3059be0402c16032823cb061281141cd and 4ef6ab1082090ae8d56da8f9487aa4c1357bb028 I have added the basic framework to detect and match mime type and file extension so users can add new types based on their preferences.

I think you should be able to extend you local `preview-tui` based on these. I am closing the PR as I would like to leave the implementation generic.

**Pull request closed without merging**